holoviews/plotting/bokeh/element.py

```diff
@@ -1919,6 +1928,29 @@ def initialize_plot(self, ranges=None, plot=None, plots=None, source=None):
 
         return plot
 
+    def _apply_hard_bounds(self, element, ranges):
+        """
+        Apply hard bounds to the x and y ranges of the plot. If xlim/ylim is set, limit the
+        initial viewable range to xlim/ylim, but allow navigation up to the abs max between
+        the data range and xlim/ylim. If dim range is set (e.g. via redim.range), enforce
+        as hard bounds.
+
+        """
+
+        def validate_bound(bound):
+            return bound if util.isfinite(bound) else None
+
+        min_extent_x, min_extent_y, max_extent_x, max_extent_y = map(
+            validate_bound, self.get_extents(element, ranges, range_type='combined', lims_as_soft_ranges=True)
+        )
+
+        def set_bounds(axis, min_extent, max_extent):
+            """Set the bounds for a given axis, using None if both extents are None or identical"""
+            self.handles[axis].bounds = None if min_extent == max_extent else (min_extent, max_extent)
+
+        set_bounds('x_range', min_extent_x, max_extent_x)
+        set_bounds('y_range', min_extent_y, max_extent_y)
```

Comment:
I just tested a heatmap and hit an error because we can't set the bounds for a FactorRange.. so I just updated this function call to 

```python
        def set_bounds(axis, min_extent, max_extent):
            """Set the bounds for a given axis, using None if both extents are None or identical"""
            try:
                self.handles[axis].bounds = None if min_extent == max_extent else (min_extent, max_extent)
            except ValueError:
                self.handles[axis].bounds = None
```

Reply:
at least for now, I think this is the best we can do with categorical axes.. see https://github.com/bokeh/bokeh/issues/13830